ffi/src/schema_visitor.rs

```diff
@@ -0,0 +1,1112 @@
+        let DataType::Variant(level4b_variant) = level4_fields[1].data_type() else {
+            panic!("Expected variant type (level 4b)");
+        };
+        let level4b_fields: Vec<_> = level4b_variant.fields().collect();
+        assert_eq!(level4b_fields.len(), 2);
+        assert_eq!(level4b_fields[0].name(), "metadata");
+        assert_eq!(
+            level4b_fields[0].data_type(),
+            &DataType::Primitive(PrimitiveType::Binary)
+        );
+        assert!(!level4b_fields[0].is_nullable());
+        assert_eq!(level4b_fields[1].name(), "value");
+        assert_eq!(
+            level4b_fields[1].data_type(),
+            &DataType::Primitive(PrimitiveType::Binary)
+        );
+        assert!(!level4b_fields[1].is_nullable());
+
+        // 4c: nested_struct: struct<final_array: ...>
+        assert!(level4_fields[2].is_nullable());
+        let DataType::Struct(level4c_struct) = level4_fields[2].data_type() else {
+            panic!("Expected struct type (level 4c)");
+        };
+        let level4c_fields: Vec<_> = level4c_struct.fields().collect();
+        assert_eq!(level4c_fields.len(), 1);
+        assert_eq!(level4c_fields[0].name(), "final_array");
+        assert!(!level4c_fields[0].is_nullable());
+
+        // 5: final_array: array<...>
+        let DataType::Array(level5_array) = level4c_fields[0].data_type() else {
+            panic!("Expected array type (level 5)");
+        };
+        assert!(!level5_array.contains_null());
+
+        // 6: array element: map<struct<coord: double>, double>
+        let DataType::Map(level6_map) = level5_array.element_type() else {
+            panic!("Expected map type (level 6)");
+        };
+
+        // 6b: map value: double
+        assert_eq!(
+            level6_map.value_type(),
+            &DataType::Primitive(PrimitiveType::Double)
+        );
+        assert!(!level6_map.value_contains_null());
+
+        // 6a: map key: struct<coord: double>
+        let DataType::Struct(level6a_key_struct) = level6_map.key_type() else {
+            panic!("Expected struct type for map key (level 6a)");
+        };
+        let level6a_key_fields: Vec<_> = level6a_key_struct.fields().collect();
+        assert_eq!(level6a_key_fields.len(), 1);
+        assert_eq!(level6a_key_fields[0].name(), "coord");
+        assert_eq!(
+            level6a_key_fields[0].data_type(),
+            &DataType::Primitive(PrimitiveType::Double)
+        );
+        assert!(!level6a_key_fields[0].is_nullable());
+    }
+
+    // // TODO: manndp review by hand (vibe-coded).
+    #[test]
+    fn test_nullability_combinations() {
+        let mut state = KernelSchemaVisitorState::default();
+
+        // Test more nullability cases:
+        // Schema:
+        // struct<
+        //   col_required_string: string NOT NULL,
+        //   col_nullable_string: string NULL,
+        //   col_nullable_array_non_null_elements: array<string> NULL (elements NOT NULL),
```

Comment:
If elements are not null, it should like like this, no? Also, maybe we can omit the explicit NULL?
Also, I feel like the "(elements NOT NULL)" comments more confusing.
```suggestion
        //   col_nullable_array_non_null_elements: array<string NOT NULL>,
```